# Numpy100 Exercise
[これ](https://github.com/rougier/numpy-100/blob/master/100_Numpy_exercises.ipynb)をやってく.
先駆者の方のは[こちら](https://github.com/osawat/numpy-100/blob/master/100_Numpy_exercises.ja.ipynb)

悩んで分からず答え見た奴は所感と考察を絶対述べる縛りで行きます.(大体わかってないはず)  
調べた場合はそのリンクを載せていく.

## 1.numpy パッケージをnpとして読み込む（難易度１）

In [1]:
import numpy as np

## 2.numpyのバージョンを確認する（難易度１）

In [4]:
# これは知ってた
print(np.__version__)

1.17.2


In [22]:
# これは知らなかった、答え合わせの時に初めて知った
np.show_config()

blas_mkl_info:
    libraries = ['blas', 'cblas', 'lapack', 'pthread', 'blas', 'cblas', 'lapack']
    library_dirs = ['/Users/Zi/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/Zi/anaconda3/include']
blas_opt_info:
    libraries = ['blas', 'cblas', 'lapack', 'pthread', 'blas', 'cblas', 'lapack', 'blas', 'cblas', 'lapack']
    library_dirs = ['/Users/Zi/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/Zi/anaconda3/include']
lapack_mkl_info:
    libraries = ['blas', 'cblas', 'lapack', 'pthread', 'blas', 'cblas', 'lapack']
    library_dirs = ['/Users/Zi/anaconda3/lib']
    define_macros = [('SCIPY_MKL_H', None), ('HAVE_CBLAS', None)]
    include_dirs = ['/Users/Zi/anaconda3/include']
lapack_opt_info:
    libraries = ['blas', 'cblas', 'lapack', 'pthread', 'blas', 'cblas', 'lapack', 'blas', 'cblas', 'lapack']
    library_dirs = ['/Users/Zi/anaconda3/lib']
    define_macr

#### 所感・考察・調査
何かblasとかcblasとかって書いてあるものに関しては[ここ](http://azalea.s35.xrea.com/blas/whatIsBlas.html)に色々書いてあった.  
numpyの裏で動いている関数群のこと,それぞれpipかcondaかで速度の違いもある模様.参考は[こちら](https://orizuru.io/blog/machine-learning/anaconda_pip_difference/).

## 3. サイズ10の零ベクトルを作成する（難易度１）

In [14]:
# 所感：元のが"null vector"ってなっててnullのベクトル？？ってなってなかなか悩んだ普通に零ベクトルのことだった
a=np.zeros(10)

In [15]:
a

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

## 4. 行列のメモリサイズを確認するには？（難易度１）

In [20]:
from sys import getsizeof
print(a.size*a.itemsize)
print(a.nbytes)
print(getsizeof(a))

80
80
176


#### 所感・考察・調査
`size` とかは割と使うので行けたが `itemsize` が思い出せなかった.  
ちなみにそれぞれの挙動は以下のよう.

In [17]:
print(a.size)
print(a.itemsize)

10
8


そもそも行列のメモリサイズが`size*itemsize`で表されることを認識していなかった.  
いつもは`nbytes`で確認していた.

`getsizeof`(176)と`nbytes`(80)の値が違うのは

- nbytes 格納されているデータのサイズのみを返す
- getsizeof 配列の次元,データ型などの情報も含むオブジェクト全体のサイズを返す

ためである.らしい.[参照元](https://living-sun.com/ja/python/727350-nbytes-and-getsizeof-return-different-values-python-numpy-memory.html)

## 5. コマンドラインから`numpy`の`add`ファンクションのドキュメントを得るには？（難易度１）

In [51]:
# なぜかマジックコマンドでできなかった…
%run python -c "import numpy; numpy.info(numpy.add)"

ERROR:root:File `'python.py'` not found.


In [52]:
!python -c "import numpy; numpy.info(numpy.add)"

add(x1, x2, /, out=None, *, where=True, casting='same_kind', order='K', dtype=None, subok=True[, signature, extobj])

Add arguments element-wise.

Parameters
----------
x1, x2 : array_like
    The arrays to be added. If ``x1.shape != x2.shape``, they must be broadcastable to a common shape (which becomes the shape of the output).
out : ndarray, None, or tuple of ndarray and None, optional
    A location into which the result is stored. If provided, it must have
    a shape that the inputs broadcast to. If not provided or `None`,
    a freshly-allocated array is returned. A tuple (possible only as a
    keyword argument) must have length equal to the number of outputs.
where : array_like, optional
    This condition is broadcast over the input. At locations where the
    condition is True, the `out` array will be set to the ufunc result.
    Elsewhere, the `out` array will retain its original value.
    Note that if an uninitialized `out` array is created via the default
    ``out=None`

#### 所感・考察・調査
マジックコマンドが動かなかったのが謎,Ipython系使ってるなら以下で良さそう.  
(このexerciseの主旨からはズレるかもだが…)

`?np.add`

IPython系での`%`,`!`の違いに関しては[こちら](https://qiita.com/2-propanol/items/f6f63c3f20981df2c844)を参考.